### PR TITLE
Add mitigations for security vulnerabilities

### PR DESCRIPTION
**Pull request by PawelPeczek-Roboflow**
# Description
This adds mitigation for sec. vulnerabilities that were marked as critical:
* server request forgery (which we can only partially prevent)
* numpy deserialisation

Scope of change:
* numpy deserialisation disabled by default (people can enable if they want) - (breaking for some clients - decision makers acknowledged)
* there are new options to:
  * disable URL inputs at all (get rid of vulnerability)
  * disable URLs without FQDN (breaking for some clients - decision makers acknowledged) 
  * disable non-https URLs (breaking for some clients - decision makers acknowledged) 
  * ability to specify whitelist of destinations (IP / domains) or blacklist ones

## Type of change

Please delete options that are not relevant.

-   [x] Bug fix (non-breaking change which fixes an issue)
-   [x] New feature (non-breaking change which adds functionality)
-   [x] This change requires a documentation update

## How has this change been tested, please provide a testcase or example of how you tested the change?

* new tests in CI
* tested local server e2e
* tested at staging e2e

## Any specific deployment considerations

For example, documentation changes, usability, usage/costs, secrets, etc.

## Docs

-   [x] Docs updated? What were the changes: info about changes

### CODE REVIEW

**Review by grzegorz-roboflow — APPROVED** [review]

File: docs/server_configuration/accepted_input_formats.md [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+However, we also recognize the importance of having a production-ready solution. Therefore, we offer configuration 
+options that allow users to disable potentially unsafe behaviors.
+
+In this document, we explain how to configure the server to either enhance security or enable more 
+flexible behaviors, depending on your needs.
+
+
+## Deserialization of pickled `numpy` objects
+
+One of the way to make request to inference server is to send serialised numpy object:
```
Comment (yeldarby) [inline docs/server_configuration/accepted_input_formats.md:19]:
```suggestion
One of the ways to send requests to the inference server is via serialized numpy objects:
```

File: docs/server_configuration/accepted_input_formats.md [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+}
+
+res = requests.post(
+    "http://localhost:9001/infer/{task}",
+    json=infer_payload,
+)
+```
+
+Starting from version `v0.14.0`, deserialization of this type of payload is disabled by default. However, you can 
+enable it by setting environmental variable `ALLOW_NUMPY_INPUT=True`. Check [inference cli docs](../inference_helpers/inference_cli.md) to
```
Comment (yeldarby) [inline docs/server_configuration/accepted_input_formats.md:45]:
```suggestion
enable it by setting an environmental variable, `ALLOW_NUMPY_INPUT=True`. Check [inference cli docs](../inference_helpers/inference_cli.md) to
```

File: docs/server_configuration/accepted_input_formats.md [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+see how to run server with that flag. This option is **not available at Roboflow hosted inference**.
```
Comment (yeldarby) [inline docs/server_configuration/accepted_input_formats.md:46]:
```suggestion
see how to run the server with that flag. This option is **not available in Roboflow's Hosted Inference API**.
```

File: docs/server_configuration/accepted_input_formats.md [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+!!! warning
+
+    Roboflow advises all users hosting the inference server in production environments not to enable this option if 
+    there is any risk of malicious requests reaching the server.
```
Comment (yeldarby) [inline docs/server_configuration/accepted_input_formats.md:51]:
```suggestion
    the server is open to requests from the open Internet or is not locked down to accept only authenticated requests from your workspace's API key.
```

File: docs/server_configuration/accepted_input_formats.md [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    "http://localhost:9001/infer/{task}",
+    json=infer_payload,
+)
+```
+
+This option is **enabled by default**, but we recommend configuring the server to enhance security using one or more of
+the following environment variables:
+* `ALLOW_URL_INPUT` - Set to `False` to allow only HTTPS protocol in URLs. This helps ensure that domain names are 
+not maliciously resolved.
+* `ALLOW_NON_HTTPS_URL_INPUT` - set to `False` to only allow https protocol in URLs (useful to make sure domain names are
```
Comment (yeldarby) [inline docs/server_configuration/accepted_input_formats.md:81]: How is this one different from the one just above?

File: inference/core/env.py [hunk 6]
```diff
@@ -12,8 +12,25 @@
 # The project name, default is "roboflow-platform"
 PROJECT = os.getenv("PROJECT", "roboflow-platform")
 
-# Allow numpy input, default is True
-ALLOW_NUMPY_INPUT = str2bool(os.getenv("ALLOW_NUMPY_INPUT", True))
+# Allow numpy input, default is False
+ALLOW_NUMPY_INPUT = str2bool(os.getenv("ALLOW_NUMPY_INPUT", False))
+ALLOW_URL_INPUT = str2bool(os.getenv("ALLOW_URL_INPUT", True))
+ALLOW_NON_HTTPS_URL_INPUT = str2bool(os.getenv("ALLOW_NON_HTTPS_URL_INPUT", True))
+ALLOW_URL_INPUT_WITHOUT_FQDN = str2bool(os.getenv("ALLOW_URL_INPUT_WITHOUT_FQDN", True))
```
Comment (yeldarby) [inline inference/core/env.py:19]: Should we make the defaults for `ALLOW_NON_HTTPS_URL_INPUT` and `ALLOW_URL_INPUT_WITHOUT_FQDN` False?

Comment (PawelPeczek-Roboflow) [reply inference/core/env.py:19]: 👍

File: docs/server_configuration/accepted_input_formats.md (hunk 5 above)
Comment (PawelPeczek-Roboflow) [reply docs/server_configuration/accepted_input_formats.md:81]: yeah, mistake

**Review by hansent — APPROVED** [review]
discussed togehter on tuple.  ACK on no longer allowing http / non FQDN urls by default

**Pull request merged**